onnxruntime/contrib_ops/cuda/custom_layer_norm.cu

```diff
@@ -0,0 +1,305 @@
+of complete warps in the threadblock.
+For FP16, this kernel does not promote to FP32 in order to utilize the 2x throughput for
+__half2 instructions, and avoid the conversion overhead (1/8 of __hal2 arithmetic).
+For specific launch constraints, see the launch functions.
+*/
+
+template <int row_stride, int iterations>
+__global__ void custom_layer_norm(
+    float* vals,
+    const float* residual,
+    const float* gamma,
+    const float* beta,
+    float epsilon,
+    float* invvars = nullptr,
+    float* means = nullptr) {
+  constexpr int iteration_stride = row_stride / iterations;
+
+  cg::thread_block b = cg::this_thread_block();
+  cg::thread_block_tile<GPU_WARP_SIZE> g = cg::tiled_partition<GPU_WARP_SIZE>(b);
+
+  int row = blockIdx.x;
+  int id = threadIdx.x;
+  int gid = id / GPU_WARP_SIZE;
+
+  float vals_arr[iterations];
+  __shared__ float shr[iteration_stride >> GPU_WARP_SIZE_LOG2];
+
+  float sum = 0.f;
+#pragma unroll
+  for (int i = 0; i < iterations; i++) {
+    vals_arr[i] = residual[row * row_stride + i * iteration_stride + id];
+    sum += vals_arr[i];
+  }
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride);
+  }
+
+  if (g.thread_rank() == 0) shr[gid] = sum;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> GPU_WARP_SIZE_LOG2)) sum = shr[g.thread_rank()];
+
+  b.sync();
+
+  for (int stride = 1; stride < (iteration_stride >> GPU_WARP_SIZE_LOG2); stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride);
+  }
+
+  sum = WARP_SHFL(sum, 0);
+  float mean = sum / row_stride;
+  if (g.thread_rank() == 0) means[row] = mean;
+  float variance = 0.f;
+  for (int i = 0; i < iterations; i++) {
+    variance += (vals_arr[i] - mean) * (vals_arr[i] - mean);
+  }
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+
+  if (g.thread_rank() == 0) shr[gid] = variance;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> GPU_WARP_SIZE_LOG2)) variance = shr[g.thread_rank()];
+
+  b.sync();
+
+  for (int stride = 1; stride < (iteration_stride >> GPU_WARP_SIZE_LOG2); stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+  variance = WARP_SHFL(variance, 0);
+  variance /= row_stride;
+  variance += epsilon;
+  const float inv_variance = rsqrtf(variance);
+  if (g.thread_rank() == 0) invvars[row] = inv_variance;
+
+  for (int i = 0; i < iterations; i++) {
+    vals_arr[i] = (vals_arr[i] - mean) * inv_variance;
+    vals_arr[i] =
+        vals_arr[i] * gamma[i * iteration_stride + id] + beta[i * iteration_stride + id];
+    vals[row * row_stride + i * iteration_stride + id] = vals_arr[i];
+  }
+}
+
+template <int row_stride, int iterations>
+__global__ void custom_layer_norm(
+    __half* vals,
+    const __half* residual,
+    const __half* gamma,
+    const __half* beta,
+    float epsilon,
+    float* invvars = nullptr,
+    float* means = nullptr) {
+#if __CUDA_ARCH__ >= 700
```

Comment:
i realize this is in the original implementation, but do we want to just silently not do anything if this condition is not met?

Reply:
We are checking for this in the launch function